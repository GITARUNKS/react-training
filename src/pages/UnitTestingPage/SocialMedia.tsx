import { useState } from "react";

interface SocialMediaProps{
    name: string;
    followersCount: string;
}

const SocialMedia: React.FC<SocialMediaProps> = (props) => {
    const [countryName, setCountryName] = useState('India');
  return (
    <div>
        <h2 role="heading" className="text-center">Social Media Info</h2>
        <h3>Demo of Testing Props, States, Events, JSX, Styles, Snapshot testing !!!</h3>
        <p> Top social media platform</p>
        <p data-testid="followersDetails" style={{backgroundColor: '#ffff00'}}>{props.name} with {props.followersCount} followers</p>

        <input type="text" name="countryName" placeholder="Enter a country name" value={countryName} onChange={(event) => {
            setCountryName(event.target.value);
        }}/>
        <p data-testid="websiteInfo">Visit Genesys {countryName} website</p>
    </div>
  )
}

export default SocialMedia