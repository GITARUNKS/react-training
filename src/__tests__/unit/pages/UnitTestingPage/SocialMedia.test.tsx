import { render, screen, fireEvent } from "@testing-library/react";
import SocialMedia from "../../../../pages/UnitTestingPage/SocialMedia";
import renderer from "react-test-renderer";

describe('SocialMedia', () => {
    // Test : Case insenstive check test
    it('has Social Media in heading', () => {
        render(<SocialMedia name="Instagram" followersCount="2M+" />);
        const headingElement = screen.getByText(/SociaL Media Info/i);
        expect(headingElement).toBeDefined();
    })

    // Test : How to pass props in a component and test
    it('has received props name and followers', () => {
        render(<SocialMedia name="Facebook" followersCount="2M+" />);
        const followersDetailsElement = screen.getByTestId('followersDetails');
        expect(followersDetailsElement.textContent).toBe("Facebook with 2M+ followers");
    })

    // Test : Form element is present or not?
    it('has input field with country name', () => {
        render(<SocialMedia name="Instagram" followersCount="2M+" />);
        const countryInput = screen.getByPlaceholderText("Enter a country name");
        expect(countryInput).toBeTruthy();
        
    })

    // Test : How to access the input element?
    it('has input country name with default value', () => {
        render(<SocialMedia name="YouTube" followersCount="2M+" />);
        // find the element in which you have on change event
        const countryInput: HTMLInputElement = screen.getByRole("textbox");
        expect(countryInput.value).toBeTruthy();
        
    })

    // Test : Testing Events and State
    it('should update country name on change', () => {
        render(<SocialMedia name="YouTube" followersCount="2M+" />);
        const countryInput: HTMLInputElement = screen.getByRole("textbox");
        //Trigerring or emitting event through program
        fireEvent.change(countryInput, {
            target: {
                value: 'Singapore'
            }
        })
        expect(countryInput.value).toBe('Singapore');
        // Multiple expect in same block
        const websiteInfoElement = screen.getByTestId('websiteInfo');
        expect(websiteInfoElement.textContent).toBe('Visit Genesys Singapore website')
    })

    // Test : Inline element style
    it('has background color green', () => {
        render(<SocialMedia name="YouTube" followersCount="2M+" />);
        const followersDetailsElement = screen.getByTestId('followersDetails');
        expect(followersDetailsElement).toHaveStyle('background-color: #ffff00');
    })

    // Test : External css style
    it('has header text aligned to centered', () => {
        render(<SocialMedia name="YouTube" followersCount="2M+" />);
        const headingElement = screen.getByText('Social Media Info');
        expect(headingElement).toHaveClass('text-center');
    })

    // Test : Snapshot testing
    it('has right Social Media component snapshot', () => {
        
        const snapshotJson = renderer.create(<SocialMedia name="YouTube" followersCount="2M+" />);
        expect(snapshotJson).toMatchSnapshot();
    })
})



