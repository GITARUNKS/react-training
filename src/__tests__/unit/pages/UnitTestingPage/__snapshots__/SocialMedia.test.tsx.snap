// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialMedia has right Social Media component snapshot 1`] = `
<div>
  <h2
    className="text-center"
    role="heading"
  >
    Social Media Info
  </h2>
  <h3>
    Demo of Testing Props, States, Events, JSX, Styles, Snapshot testing !!!
  </h3>
  <p>
     Top social media platform
  </p>
  <p
    data-testid="followersDetails"
    style={
      {
        "backgroundColor": "#ffff00",
      }
    }
  >
    YouTube
     with 
    2M+
     followers
  </p>
  <input
    name="countryName"
    onChange={[Function]}
    placeholder="Enter a country name"
    type="text"
    value="India"
  />
  <p
    data-testid="websiteInfo"
  >
    Visit Genesys 
    India
     website
  </p>
</div>
`;
